File: binding/src/multimedia/libqscreencapture.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qscreencapture.html
pub const qscreencapture = struct {

    /// New constructs a new QScreenCapture object.
    pub fn New() QtC.QScreenCapture {
        return qtc.QScreenCapture_new();
    }


    /// New2 constructs a new QScreenCapture object.
    pub fn New2(parent: ?*anyopaque) QtC.QScreenCapture {
        return qtc.QScreenCapture_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScreenCapture_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscreencapture.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn CaptureSession(self: ?*anyopaque, ) QtC.QMediaCaptureSession {
        return qtc.QScreenCapture_CaptureSession(@ptrCast(self));
    }

    pub fn SetScreen(self: ?*anyopaque, screen: ?*anyopaque) void {
        qtc.QScreenCapture_SetScreen(@ptrCast(self), @ptrCast(screen));
    }

    pub fn Screen(self: ?*anyopaque, ) QtC.QScreen {
        return qtc.QScreenCapture_Screen(@ptrCast(self));
    }

    pub fn IsActive(self: ?*anyopaque, ) bool {
        return qtc.QScreenCapture_IsActive(@ptrCast(self));
    }

    pub fn Error(self: ?*anyopaque, ) i32 {
        return qtc.QScreenCapture_Error(@ptrCast(self));
    }

    pub fn ErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScreenCapture_ErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscreencapture.ErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetActive(self: ?*anyopaque, active: bool) void {
        qtc.QScreenCapture_SetActive(@ptrCast(self), active);
    }

    pub fn Start(self: ?*anyopaque, ) void {
        qtc.QScreenCapture_Start(@ptrCast(self));
    }

    pub fn Stop(self: ?*anyopaque, ) void {
        qtc.QScreenCapture_Stop(@ptrCast(self));
    }

    pub fn ActiveChanged(self: ?*anyopaque, param1: bool) void {
        qtc.QScreenCapture_ActiveChanged(@ptrCast(self), param1);
    }

    pub fn OnActiveChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QScreenCapture_Connect_ActiveChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ErrorChanged(self: ?*anyopaque, ) void {
        qtc.QScreenCapture_ErrorChanged(@ptrCast(self));
    }

    pub fn OnErrorChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QScreenCapture_Connect_ErrorChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ScreenChanged(self: ?*anyopaque, param1: ?*anyopaque) void {
        qtc.QScreenCapture_ScreenChanged(@ptrCast(self), @ptrCast(param1));
    }

    pub fn OnScreenChanged(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QScreenCapture_Connect_ScreenChanged(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn ErrorOccurred(self: ?*anyopaque, errorVal: i32, errorString: []const u8) void {
        const errorString_str = qtc.libqt_string{
    .len = errorString.len,
    .data = errorString.ptr,
};
qtc.QScreenCapture_ErrorOccurred(@ptrCast(self), @intCast(errorVal), errorString_str);
    }

    pub fn OnErrorOccurred(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32, [*:0]const u8) callconv(.c) void) void {
        qtc.QScreenCapture_Connect_ErrorOccurred(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScreenCapture_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscreencapture.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QScreenCapture_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qscreencapture.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QScreenCapture_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/screencapture.html#types
pub const enums = struct {
    pub const Error = enum {
        pub const NoError: i32 = 0;
        pub const InternalError: i32 = 1;
        pub const CapturingNotSupported: i32 = 2;
        pub const CaptureFailed: i32 = 4;
        pub const NotFound: i32 = 5;
    };

};

File: binding/src/multimedia/libqscreencapture.cpp
#include <QMediaCaptureSession>
#include <QObject>
#include <QScreen>
#include <QScreenCapture>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <qscreencapture.h>
#include "libqscreencapture.h"
#include "libqscreencapture.hxx"

QScreenCapture* QScreenCapture_new() {
	 return new QScreenCapture();
}

QScreenCapture* QScreenCapture_new2(QObject* parent) {
	 return new QScreenCapture(parent);
}

libqt_string QScreenCapture_Tr(const char* s) {
	QString _ret = QScreenCapture::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QMediaCaptureSession* QScreenCapture_CaptureSession(const QScreenCapture* self) {
	return self->captureSession();
}

void QScreenCapture_SetScreen(QScreenCapture* self, QScreen* screen) {
	self->setScreen(screen);
}

QScreen* QScreenCapture_Screen(const QScreenCapture* self) {
	return self->screen();
}

bool QScreenCapture_IsActive(const QScreenCapture* self) {
	return self->isActive();
}

int QScreenCapture_Error(const QScreenCapture* self) {
	return self->error();
}

libqt_string QScreenCapture_ErrorString(const QScreenCapture* self) {
	QString _ret = self->errorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QScreenCapture_SetActive(QScreenCapture* self, bool active) {
	self->setActive(active);
}

void QScreenCapture_Start(QScreenCapture* self) {
	self->start();
}

void QScreenCapture_Stop(QScreenCapture* self) {
	self->stop();
}

void QScreenCapture_ActiveChanged(QScreenCapture* self, bool param1) {
	self->activeChanged(param1);
}

void QScreenCapture_Connect_ActiveChanged(QScreenCapture* self, intptr_t slot) {
    void (*slotFunc)(QScreenCapture*, bool) = reinterpret_cast<void (*)(QScreenCapture*, bool)>(slot);
    QScreenCapture::connect(self, &QScreenCapture::activeChanged, [self, slotFunc](bool param1) {
	slotFunc(self, param1);
    });
}

void QScreenCapture_ErrorChanged(QScreenCapture* self) {
	self->errorChanged();
}

void QScreenCapture_Connect_ErrorChanged(QScreenCapture* self, intptr_t slot) {
    void (*slotFunc)(QScreenCapture*) = reinterpret_cast<void (*)(QScreenCapture*)>(slot);
    QScreenCapture::connect(self, &QScreenCapture::errorChanged, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QScreenCapture_ScreenChanged(QScreenCapture* self, QScreen* param1) {
	self->screenChanged(param1);
}

void QScreenCapture_Connect_ScreenChanged(QScreenCapture* self, intptr_t slot) {
    void (*slotFunc)(QScreenCapture*, QScreen*) = reinterpret_cast<void (*)(QScreenCapture*, QScreen*)>(slot);
    QScreenCapture::connect(self, &QScreenCapture::screenChanged, [self, slotFunc](QScreen* param1) {
	slotFunc(self, param1);
    });
}

void QScreenCapture_ErrorOccurred(QScreenCapture* self, int errorVal, const libqt_string errorString) {
	self->errorOccurred(static_cast<QScreenCapture::Error>(errorVal), QString::fromUtf8(errorString.data, errorString.len));
}

void QScreenCapture_Connect_ErrorOccurred(QScreenCapture* self, intptr_t slot) {
    void (*slotFunc)(QScreenCapture*, int, const char*) = reinterpret_cast<void (*)(QScreenCapture*, int, const char*)>(slot);
    QScreenCapture::connect(self, &QScreenCapture::errorOccurred, [self, slotFunc](QScreenCapture::Error errorVal, const QString& errorString) {
	slotFunc(self, errorVal, errorString);
    });
}

libqt_string QScreenCapture_Tr2(const char* s, const char* c) {
	QString _ret = QScreenCapture::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QScreenCapture_Tr3(const char* s, const char* c, int n) {
	QString _ret = QScreenCapture::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QScreenCapture_Delete(QScreenCapture* self) {
    delete self;
}

